src/main.rs

```diff
@@ -9,10 +9,81 @@ use handlebars::Handlebars;
+ * The handlebars templates can use any information found in the codegen object.
+ * The formatted output is written to the file in output_filename.
+ * Arguments:
+ * @code_gen:  a code_gen object that contains information that can be formatted nicely by the handlebars
+ * @template_path_name: the path leading to a handlebars template
+ * @output_filename: where the output is written
+ */
+fn generate_code_from_codegen_with_handlebars(code_gen: &code_gen::CodeGen, template_path: PathBuf, output_filename: PathBuf) {
+    let display = template_path.display();
+    let mut template_file = match File::open(&template_path) {
+        Err(msg) => panic!("Failed to open {}: {}", display, msg),
+        Ok(file) => file,
+    };
+
+    let mut template_str = String::new();
+    match template_file.read_to_string(&mut template_str) {
+        Err(msg) => panic!("Failed to read {}: {}", display, msg),
+        Ok(_) => println!("Successfully read {}", display),
+    }
+
+    let handlebars = Handlebars::new();
+
+    let output = handlebars
+        .render_template(&template_str, &code_gen)
+        .expect("handlebar render failed");
+
+    let mut file = File::create(output_filename).expect("file create failed.");
+    file.write_all(output.as_bytes()).expect("write failed");
+}
+
+/* Generates code using templates formatted by handlebars
+ * (see https://docs.rs/handlebars/3.5.2/handlebars/)
+ * The handlebars templates are given filter_name as input;  it is their only piece of information.
+ * The formatted output is written to the file in output_filename.
+ * Arguments:
+ * @filter_name:  the name of the filter;  also the only information used by the handlebars formatter
+ * @template_path_name: the path leading to a handlebars template
+ * @output_filename: where the output is written
+ */
+fn generate_code_from_filter_name_with_handlebars(filter_name: String, template_path: PathBuf, output_filename: PathBuf) {
+    let display = template_path.display();
+    let mut template_file = match File::open(&template_path) {
+        Err(msg) => panic!("Failed to open {}: {}", display, msg),
+        Ok(file) => file,
+    };
+
+    let mut template_str = String::new();
+    match template_file.read_to_string(&mut template_str) {
+        Err(msg) => panic!("Failed to read {}: {}", display, msg),
+        Ok(_) => println!("Successfully read {}", display),
+    }
+
+    let handlebars = Handlebars::new();
+
+    let mut data = BTreeMap::new();
+    data.insert("filter_name".to_string(), filter_name);
+    let output = handlebars
+        .render_template(&template_str, &data)
+        .expect("handlebar render failed");
+
+    let mut file = File::create(output_filename).expect("file create failed.");
+    file.write_all(output.as_bytes()).expect("write failed");
+}
+
+
 fn main() {
-    let bin_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
-    let def_filter_dir = bin_dir.join("cpp_filter/filter.cc");
```

Comment:
Do not remove this. It is needed to make the compiler binary independent of the place you are calling it from.

Reply:
Oh, I see.  I thought those variables were never used but you used them to create the filter.cc file.  Is it necessary to prepend that to every file we create, including the rust library?

Reply:
Yes, especially the rust library. Essentially, it makes any path independent from the execution point because we define it relative to the location of the toml dir.

Reply:
Sounds good.  All rust code produced now has that before it.